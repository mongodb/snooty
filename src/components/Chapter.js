import React, { useMemo } from 'react';
import { withPrefix } from 'gatsby';
import styled from '@emotion/styled';
import { uiColors } from '@leafygreen-ui/palette';
import Link from './Link';
import { getPlaintext } from '../utils/get-plaintext';

// Height and width of iamge
const IMAGE_SIZE = 200;

const Container = styled('div')`
  background-color: ${uiColors.white};
  border-radius: 4px;
  border: 1px solid ${uiColors.gray.light3};
  display: grid;
  grid-template-areas:
    'description image'
    'guides guides';
  margin: 8px 0;
  max-width: 743px;
  padding: 48px 56px;
`;

const ImageContainer = styled('img')`
  grid-area: image;
  margin-left: 108px;
`;

const DescriptionContainer = styled('div')`
  display: flex;
  flex-direction: column;
  grid-area: description;
  justify-content: center;
`;

const ChapterNumberLabel = styled('div')`
  background-color: ${uiColors.green.light3};
  border-radius: 4px;
  color: ${uiColors.green.base};
  font-size: 14px;
  font-weight: bold;
  height: 25px;
  text-align: center;
  width: 83px;
`;

const ChapterTitle = styled('div')`
  font-size: 18px;
  font-weight: bold;
  margin-top: 8px;
`;

const ChapterDescription = styled('div')`
  margin-top: 8px;
`;

const GuidesList = styled('ul')`
  grid-area: guides;
  list-style-type: none;
  list-style-image: url(${withPrefix('assets/lightning-bolt.svg')});
  margin-top: 24px;
  margin-bottom: 0;
  padding-inline-start: 14px;
`;

const GuidesListItem = styled('li')`
  padding-top: 4px;
  padding-bottom: 4px;
`;

const GuideLink = styled(Link)`
  align-items: center;
  display: flex;
  color: unset;
  justify-content: space-between;
  position: relative;

  :hover {
    color: unset;
  }
`;

// Uses guides metadata to obtain the title and completion time of each guide in the chapter
const getGuidesData = (metadata, chapterTitle) => {
  const guidePaths = metadata?.chapters[chapterTitle]?.guides;
  let guides = [];

  if (guidePaths) {
    guidePaths.forEach((guidePath) => {
      const guide = metadata?.guides?.[guidePath];
      guides.push({
        path: guidePath,
        time: guide?.completion_time,
        title: guide?.title,
      });
    });
  }

  return guides;
};

const Chapter = ({ metadata, nodeData: { argument, options } }) => {
  const description = options?.description;
  const image = options?.image;
  const chapterTitle = getPlaintext(argument);
  const chapterNumber = metadata?.chapters?.[chapterTitle]?.chapter_number;
  const guides = useMemo(() => getGuidesData(metadata, chapterTitle), [metadata, chapterTitle]);

  return (
    <Container>
      <DescriptionContainer>
        <ChapterNumberLabel>Chapter {chapterNumber}</ChapterNumberLabel>
        <ChapterTitle>{chapterTitle}</ChapterTitle>
        <ChapterDescription>{description}</ChapterDescription>
      </DescriptionContainer>
      {image && (
        <ImageContainer src={withPrefix(image)} altText="Chapter image" height={IMAGE_SIZE} width={IMAGE_SIZE} />
      )}
      <GuidesList>
        {guides.map((guide) => {
          const time = guide.time ? `${guide.time} mins` : '';
          const guideTitle = getPlaintext(guide.title);

          return (
            <GuidesListItem>
              <GuideLink to={guide.path}>
                <span>{guideTitle}</span>
                <span>{time}</span>
              </GuideLink>
            </GuidesListItem>
          );
        })}
      </GuidesList>
    </Container>
  );
};

export default Chapter;
