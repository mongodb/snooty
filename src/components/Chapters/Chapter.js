import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { withPrefix } from 'gatsby';
import styled from '@emotion/styled';
import Card from '@leafygreen-ui/card';
import { palette } from '@leafygreen-ui/palette';
import { Body } from '@leafygreen-ui/typography';
import ChapterNumberLabel from './ChapterNumberLabel';
import Link from '../Link';
import { theme } from '../../theme/docsTheme';
import { getPlaintext } from '../../utils/get-plaintext';

// Height and width of image
const IMAGE_SIZE = 200;

const Container = styled(Card)`
  background-color: ${palette.white};
  border: 1px solid ${palette.gray.light3};
  padding: ${theme.size.large} ${theme.size.medium};

  @media ${theme.screenSize.mediumAndUp} {
    display: grid;
    grid-template-areas:
      'description image'
      'guides guides';
    padding: 48px 40px;
    grid-column-gap: ${theme.size.large};
  }

  @media ${theme.screenSize.largeAndUp} {
    grid-column-gap: 40px;
    scroll-margin-top: 180px;
  }

  @media ${theme.screenSize['2XLargeAndUp']} {
    grid-column-gap: 50px;
  }

  @media ${theme.screenSize['3XLargeAndUp']} {
    grid-column-gap: 108px;
  }
`;

const ChapterImage = styled('img')`
  display: block;
  margin: 0 auto ${theme.size.medium} auto;
  height: auto;
  max-width: 100%;

  @media ${theme.screenSize.mediumAndUp} {
    grid-area: image;
    margin-bottom: 0;
  }
`;

// Reserves enough space for chapters without images, while keeping existing images for chapters responsive
const EmptyImage = styled('div')`
  height: auto;
  width: ${IMAGE_SIZE}px;

  @media ${theme.screenSize.mediumAndUp} {
    grid-area: image;
  }
`;

const DescriptionContainer = styled('div')`
  display: flex;
  flex-direction: column;
  justify-content: center;

  @media ${theme.screenSize.mediumAndUp} {
    grid-area: description;
  }
`;

const ChapterTitle = styled('div')`
  font-size: 18px;
  font-weight: bold;
  margin-top: ${theme.size.small};
`;

const ChapterDescription = styled(Body)`
  margin-top: ${theme.size.small};
`;

const GuidesList = styled('ul')`
  list-style-type: none;
  list-style-image: url(${withPrefix('assets/lightning-bolt.svg')});
  margin-bottom: 0;
  margin-top: ${theme.size.medium};
  padding-inline-start: 14px;

  @media ${theme.screenSize.mediumAndUp} {
    grid-area: guides;
  }
`;

const GuideLink = styled(Link)`
  border-radius: ${theme.size.tiny};
  display: flex;
  color: unset;
  flex-direction: column;
  padding: ${theme.size.small};
  position: relative;
  font-size: ${theme.size.default};
  line-height: ${theme.size.medium};

  :hover {
    background-color: ${palette.gray.light2};
    color: unset;

    ::after {
      content: none;
    }
  }

  @media ${theme.screenSize.mediumAndUp} {
    align-items: center;
    flex-direction: row;
    justify-content: space-between;
    padding: ${theme.size.tiny} ${theme.size.small};
  }
`;

// Uses guides metadata to obtain the title and completion time of each guide in the chapter
const getGuidesData = (metadata, currentChapter) => {
  const guidePaths = currentChapter?.guides;
  let guides = [];

  if (guidePaths) {
    guidePaths.forEach((guidePath) => {
      const guide = metadata?.guides?.[guidePath];
      guides.push({
        path: guidePath,
        time: guide?.completion_time,
        title: guide?.title,
      });
    });
  }

  return guides;
};

const Chapter = ({ metadata, nodeData: { argument, options } }) => {
  const description = options?.description;
  const image = options?.image;
  const chapterTitle = getPlaintext(argument);
  const currentChapter = metadata?.chapters?.[chapterTitle];
  const chapterNumber = currentChapter?.chapter_number;
  const guides = useMemo(() => getGuidesData(metadata, currentChapter), [metadata, currentChapter]);

  return (
    // TODO: have to return specific typography here. card is explicit size 13 font

    <Container id={currentChapter?.id}>
      {image ? (
        <ChapterImage
          src={withPrefix(image)}
          height={IMAGE_SIZE}
          width={IMAGE_SIZE}
          altText={`Chapter image for ${chapterTitle}`}
        />
      ) : (
        <EmptyImage />
      )}
      <DescriptionContainer>
        <ChapterNumberLabel number={chapterNumber} />
        <ChapterTitle>{chapterTitle}</ChapterTitle>
        <ChapterDescription>{description}</ChapterDescription>
      </DescriptionContainer>
      <GuidesList>
        {guides.map((guide, i) => {
          const time = guide.time ? `${guide.time} mins` : '';
          const guideTitle = getPlaintext(guide.title);

          return (
            <li key={`${guideTitle}-${i}`}>
              <GuideLink to={guide.path}>
                <span>{guideTitle}</span>
                <span>{time}</span>
              </GuideLink>
            </li>
          );
        })}
      </GuidesList>
    </Container>
  );
};

Chapter.propTypes = {
  metadata: PropTypes.shape({
    chapters: PropTypes.object.isRequired,
    guides: PropTypes.object.isRequired,
  }).isRequired,
  nodeData: PropTypes.shape({
    argument: PropTypes.arrayOf(PropTypes.object).isRequired,
    options: PropTypes.shape({
      description: PropTypes.string,
      image: PropTypes.string,
    }),
  }).isRequired,
};

export default Chapter;
