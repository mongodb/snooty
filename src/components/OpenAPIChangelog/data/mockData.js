export const mockChangelog = [
  {
    date: '2024-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'removed',
            changes: [
              {
                change: 'resource version 2023-01-01 was removed, latest available resource version is 2023-02-01',
                changeCode: 'resource-version-removed',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
              {
                change:
                  "for the 'query' request parameter 'clusterName', default value was changed from 'default' to 'cluster0'",
                changeCode: 'request-parameter-default-value-changed',
                backwardCompatible: false,
              },
              {
                change: "added the optional property 'links' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'deprecate',
            sunsetDate: '2024-02-01',
            changes: [
              {
                change: 'deprecated by resource version 2023-02-01 and marked for removal for 2024-02-01.',
                changeCode: 'resource-version-deprecated',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/accessList',
        httpMethod: 'POST',
        operationId: 'createProjectIpAccessList',
        tag: 'Project IP Access List',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change: "added the optional property 'name' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
              {
                change: "added the optional property 'comment' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-22',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change:
                  "added the optional property 'withDefaultAlertsSettings' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
];

export const mockDiff = [
  {
    path: '/api/atlas/v2/federationSettings/{federationSettingsId}/connectedOrgConfigs',
    httpMethod: 'GET',
    operationId: 'listConnectedOrgConfigs',
    tag: 'Federated Authentication',
    changeType: 'update',
    changes: [
      {
        change: "removed the non-success response with the status '400'.",
        changeCode: 'response-non-success-status-removed',
        backwardCompatible: false,
      },
      {
        change: "added the new optional 'query' request parameter 'envelope'.",
        changeCode: 'new-optional-request-parameter',
        backwardCompatible: true,
      },
    ],
  },
  {
    path: '/api/atlas/v2/groups/{groupId}/accessList',
    httpMethod: 'POST',
    operationId: 'createProjectIpAccessList',
    tag: 'Project IP Access List',
    changeType: 'removed',
    changes: [
      {
        change: "removed the required property 'name' from the response with the '200' status.",
        changeCode: 'response-required-property-removed',
        backwardCompatible: false,
      },
    ],
  },
];

export const mockIndex = {
  runDate: '2023-04-03',
  runId: '20230403105948111',
  specRevision: '422ab47f864909549362a7b39688404c82e2540b',
  specRevisionShort: '422ab47f864',
  versions: ['2023-01-01', '2023-02-01', '2023-08-11~preview'],
};
