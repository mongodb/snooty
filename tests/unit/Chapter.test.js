import React from 'react';
import { mount } from 'enzyme';
import mockData from './data/Chapters.test.json';
import Chapter from '../../src/components/Chapters/Chapter';

it('renders correctly', () => {
  const chapterData = mockData.nodeData.children[0];
  const wrapper = mount(<Chapter nodeData={chapterData} metadata={mockData.metadata} />);
  expect(wrapper).toMatchSnapshot();

  // Make sure that the logic used to get the component's rendered data is correct
  expect(wrapper.find('ChapterNumberLabel').text()).toEqual('Chapter 1');
  expect(wrapper.find('ChapterTitle').text()).toEqual('Atlas');
  expect(wrapper.find('GuidesListItem')).toHaveLength(3);
});
