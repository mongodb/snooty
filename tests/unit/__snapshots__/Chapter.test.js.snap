// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  background-color: #FFFFFF;
  border-radius: 4px;
  border: 1px solid #F9FBFA;
  padding: 32px 24px;
}

@media not all and (max-width: 767px) {
  .emotion-0 {
    display: grid;
    grid-template-areas: 'description image' 'guides guides';
    padding: 48px 40px;
    grid-column-gap: 32px;
  }
}

@media not all and (max-width: 1023px) {
  .emotion-0 {
    grid-column-gap: 40px;
    scroll-margin-top: 180px;
  }
}

@media not all and (max-width: 1440px) {
  .emotion-0 {
    grid-column-gap: 50px;
  }
}

@media not all and (max-width: 1920px) {
  .emotion-0 {
    grid-column-gap: 108px;
  }
}

.emotion-2 {
  position: relative;
  -webkit-transition: 150ms ease-in-out;
  transition: 150ms ease-in-out;
  transition-property: border,box-shadow;
  border: 1px solid #E8EDEB;
  box-shadow: 0 4px 10px -4px rgba(0,30,43,0.3);
  background-color: white;
  color: #1C2D38;
  border-radius: 24px;
  font-family: 'Euclid Circular A',Akzidenz,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 13px;
  line-height: 20px;
  padding: 24px;
  min-height: 68px;
}

.emotion-3 {
  display: block;
  margin: 0 auto 24px auto;
  height: auto;
  max-width: 100%;
}

@media not all and (max-width: 767px) {
  .emotion-3 {
    grid-area: image;
    margin-bottom: 0;
  }
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

@media not all and (max-width: 767px) {
  .emotion-5 {
    grid-area: description;
  }
}

.emotion-7 {
  background-color: #E4F4E4;
  border-radius: 4px;
  color: #13AA52;
  font-size: 14px;
  font-weight: bold;
  height: 24px;
  text-align: center;
  width: 83px;
}

.emotion-9 {
  font-size: 18px;
  font-weight: bold;
  margin-top: 8px;
}

.emotion-11 {
  margin-top: 8px;
}

.emotion-13 {
  list-style-type: none;
  list-style-image: url(/assets/lightning-bolt.svg);
  margin-bottom: 0;
  margin-top: 24px;
  -webkit-padding-start: 14px;
  padding-inline-start: 14px;
}

@media not all and (max-width: 767px) {
  .emotion-13 {
    grid-area: guides;
  }
}

.emotion-15 {
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: unset;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 8px;
  position: relative;
}

.emotion-15:hover {
  background-color: #E7EEEC;
  color: unset;
  -webkit-text-decoration: none;
  text-decoration: none;
}

@media not all and (max-width: 767px) {
  .emotion-15 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
    padding: 4px 8px;
  }
}

<div
    class="emotion-0 emotion-1 emotion-2"
    id="atlas"
  >
    <img
      class="emotion-3 emotion-4"
      height="200"
      src="/images/chapter-image-atlas.png"
      width="200"
    />
    <div
      class="emotion-5 emotion-6"
    >
      <div
        class="emotion-7 emotion-8"
      >
        Chapter 1
      </div>
      <div
        class="emotion-9 emotion-10"
      >
        Atlas
      </div>
      <div
        class="emotion-11 emotion-12"
      >
        Get up and running with MongoDB Atlas. Learn how to create a cluster, load a sample dataset, and get your connection information.
      </div>
    </div>
    <ul
      class="emotion-13 emotion-14"
    >
      <li>
        <a
          class="emotion-15 emotion-16"
          href="/cloud/account/"
        >
          <span>
            Create and Manage a MongoDB Account
          </span>
          <span>
            15 mins
          </span>
        </a>
      </li>
      <li>
        <a
          class="emotion-15 emotion-16"
          href="/cloud/connectionstring/"
        >
          <span>
            Set Up Atlas Connectivity
          </span>
          <span />
        </a>
      </li>
      <li>
        <a
          class="emotion-15 emotion-16"
          href="/cloud/migrate-from-aws-to-atlas/"
        >
          <span>
            Migrate a MongoDB Replicate Set from AWS to MongoDB Atlas
          </span>
          <span />
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
